File: sharedUI/src/commonMain/kotlin/br/com/ajudafio/presentation/book/list/BookListViewModel.kt
package br.com.ajudafio.presentation.book.list

import androidx.lifecycle.ViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update

class BookListViewModel(
//    private val searchBooks: SearchBooksUseCase,
) : ViewModel() {

    private val _state = MutableStateFlow(
        BookListState(
            searchResults = books,
        )
    )
    val state: StateFlow<BookListState> = _state.asStateFlow()

    fun onAction(action: BookListAction) {
        when (action) {
            is BookListAction.OnBookClick -> {
                TODO()
            }
            is BookListAction.OnSearchQueryChange -> {
                _state.update {
                    it.copy(searchQuery = action.query)
                }
            }
            is BookListAction.onTabSelected -> {
                _state.update {
                    it.copy(selectedTabIndex = action.index)
                }
            }
        }
    }

}

File: iosApp/iosApp/Book/Presentation/List/BookListViewModel.swift
import Foundation
import SharedLogic

// Espelha br.com.ajudafio.presentation.book.list.BookListViewModel do sharedUI:
// mesmo state/action, mas debounce via Task.sleep no lugar de Flow.debounce,
// já que não temos Kotlin Coroutines/Flow do lado Swift.
@MainActor
final class BookListViewModel: ObservableObject {
    @Published private(set) var state = BookListState()

    private var cachedBooks: [Book] = []
    private var searchTask: Task<Void, Never>?

    func onAction(_ action: BookListAction) {
        switch action {
        case .onSearchQueryChange(let query):
            state.searchQuery = query
            scheduleSearch(query: query)

        case .onTabSelected(let index):
            state.selectedTabIndex = index

        case .onBookClick:
            break // navegação é responsabilidade da view host
        }
    }

    private func scheduleSearch(query: String) {
        searchTask?.cancel()
        searchTask = Task { [weak self] in
            try? await Task.sleep(nanoseconds: 500_000_000)
            guard let self, !Task.isCancelled else { return }
            await self.runSearch(query: query)
        }
    }

    private func runSearch(query: String) async {
        if query.isEmpty {
            state.errorMessage = nil
            state.searchResults = cachedBooks
            return
        }
        guard query.count >= 2 else { return }

        state.isLoading = true
        do {
            let results = try await BookGateway.shared.searchBooks(query: query)
            state.isLoading = false
            state.errorMessage = nil
            state.searchResults = results
        } catch let error as BookLoadException {
            state.isLoading = false
            state.searchResults = []
            state.errorMessage = error.message
        } catch {
            state.isLoading = false
            state.searchResults = []
            state.errorMessage = "Algo deu errado. Tente novamente."
        }
    }
}
